src/D2L.CodeStyle.Analyzers/Language/RequireNamedArgumentsAnalyzer.cs

```diff
@@ -123,34 +123,80 @@ SyntaxNodeAnalysisContext ctx
-			}
-
-			// Literal arguments should always be named
-			foreach( var arg in unnamedArgs ) {
-
-				// Check if the argument type is literal
-				if( arg.Syntax.Expression is LiteralExpressionSyntax ) {
-
+			}
+
+			int numOfSwappableArgs = 0;
+			foreach( var arg in unnamedArgs ) {
+
+				// Literal expressions have their own issue
+				if( arg.Syntax.Expression is LiteralExpressionSyntax ) {
+					// Literal arguments should always be named
 					var fixerContext = CreateFixerContext( ImmutableArray.Create( arg ) );
-
 					ctx.ReportDiagnostic( Diagnostic.Create(
 							descriptor: Diagnostics.LiteralArgShouldBeNamed,
 							location: arg.Syntax.Expression.GetLocation(),
 							properties: fixerContext,
-							messageArgs: arg.ParamName
-						)
-					);
+							messageArgs: arg.ParamName ) );
+					continue;
+				}
+
+				// Take note of the type of argument
+				ITypeSymbol argType = GetArgumentType( ctx, arg.Syntax );
+
+				// Iterate through the other arguments in the list
+				foreach( var compareArg in unnamedArgs ) {
+					if( compareArg.Position == arg.Position ) {
+						continue;
+					}
+
+					// Get the associated parameter of the argument to compare
+					IParameterSymbol param = compareArg.Syntax.DetermineParameter(
+						ctx.SemanticModel,
+						allowParams: false );
+
+					// Check if the argument could apply to the other parameter
+					if( ctx.SemanticModel.Compilation.ClassifyConversion(
+						argType,
+						param.Type ).Exists ) {
+						// The argument can apply to multiple parameters,
+						// so take note of it
+						numOfSwappableArgs++;
+						break;
+					}
 				}
-			}
-
-			// TODO: if there are duplicate typed args then they should be named
-			// These will create a bit more cleanup. Fix should probably name
-			// all the args instead to avoid craziness with overloading.
+			}
+
+			// If more than one argument can apply to multiple parameters,
+			// then require the arguments to be named
+			if( numOfSwappableArgs > 1 ) {
+				var fixerContext = CreateFixerContext( unnamedArgs );
+
+				ctx.ReportDiagnostic( Diagnostic.Create(
+						descriptor: Diagnostics.NamedArgumentsRequired,
+						location: ctx.Node.GetLocation(),
+						properties: fixerContext ) );
+			}
+
+			// TODO: if there are duplicate typed args then they should be named
+			// These will create a bit more cleanup. Fix should probably name
+			// all the args instead to avoid craziness with overloading.
 		}
 
-		private static ImmutableDictionary<string, string> CreateFixerContext( ImmutableArray<ArgParamBinding> unnamedArgs ) {
+		private static ITypeSymbol GetArgumentType(
+			SyntaxNodeAnalysisContext ctx,
+			ArgumentSyntax argSyntax ) {
+			var baseSymbol = ctx.SemanticModel.GetSymbolInfo( argSyntax.Expression ).Symbol;
+			return baseSymbol switch {
+				IParameterSymbol symbol => symbol.Type,
+				IFieldSymbol symbol => symbol.Type,
+				ILocalSymbol symbol => symbol.Type,
+				_ => baseSymbol.ContainingType,

```

Comment:
I don't think this is right.... I think you want `ctx.SemanticModel.GetTypeInfo` instead, without this switch.
You can pass an expression like `2 + 2` as an argument... and I think this would crash?